File: payola/s2js/runtime/src/main/scala/s2js/runtime/s2js/package.scala
package s2js.runtime.s2js

import s2js.compiler.NativeJs
import s2js.adapters.goog
import s2js.runtime.scala.NotImplementedException

object `package`
{
    @NativeJs("""
        if (!s2js.isUndefined(anObject.__class__)) {
            return anObject.__class__;
        }
        return null;
    """)
    def classOf(anObject: Any): Class = null

    private def isInstanceOf(anObject: Any, classFullName: String): Boolean = {
        val classNameIsAny = classFullName == "Any"
        val classNameIsAnyOrAnyVal = classNameIsAny || classFullName == "AnyVal"
        val classNameIsAnyOrAnyRef = classNameIsAny || classFullName == "AnyRef"
        goog.typeOf(anObject) match {
            case "undefined" | "null" => false
            case "number" => {
                classFullName match {
                    case "scala.Byte" | "scala.Short" | "scala.Int" | "scala.Long" => isInteger(anObject)
                    case "scala.Float" | "scala.Double" => true
                    case _ => classNameIsAnyOrAnyVal
                }
            }
            case "boolean" => classNameIsAnyOrAnyVal || classFullName == "scala.Boolean"
            case "string" => {
                classFullName match {
                    case "scala.Char" => isChar(anObject)
                    case "scala.String" => true
                    case _ => classNameIsAnyOrAnyRef
                }
            }
            case "function" => throw new NotImplementedException("Type check of a function isn't supported.")
            case "object" if classNameIsAnyOrAnyRef => true
            case _ if classOf(anObject) != null => classOf(anObject).isSubClassOrEqual(classFullName)
            case _ => throw new RuntimeException("Can't determine the type of the object '" + anObject.toString + ".")
        }
    }

    private def asInstanceOf(anObject: Any, className: String): Any = {
        // Just check if the conversion is possible. Nothing has to be done with the object.
        if (!isInstanceOf(anObject, className)) {
            throw new ClassCastException("The object '" + anObject.toString + "' can't be casted to " + className + ".")
        }
        anObject
    }

    @NativeJs("return goog.typeOf(anObject) === 'undefined';")
    def isUndefined(anObject: Any): Boolean = false

    @NativeJs("return anObject % 1 === 0;")
    private def isInteger(anObject: Any): Boolean = false

    @NativeJs("return anObject.length === 1;")
    private def isChar(anObject: Any): Boolean = false
}

File: payola/web/server/public/javascripts/runtime/s2js/runtime/s2js/package.js
goog.provide('s2js');
goog.require('goog');
goog.require('scala.ClassCastException');
goog.require('scala.NotImplementedException');
s2js.classOf = function(anObject) {
var self = this;

        if (!s2js.isUndefined(anObject.__class__)) {
            return anObject.__class__;
        }
        return null;
    };
s2js.isInstanceOf = function(anObject, classFullName) {
var self = this;
var classNameIsAny = (classFullName == 'Any');
var classNameIsAnyOrAnyVal = (classNameIsAny || (classFullName == 'AnyVal'));
var classNameIsAnyOrAnyRef = (classNameIsAny || (classFullName == 'AnyRef'));
return (function($selector$1) {
if (($selector$1 === 'undefined') || ($selector$1 === 'null')) {
return false;
}
if ($selector$1 === 'number') {
return (function($selector$2) {
if (($selector$2 === 'scala.Byte') || ($selector$2 === 'scala.Short') || ($selector$2 === 'scala.Int') || ($selector$2 === 'scala.Long')) {
return self.isInteger(anObject);
}
if (($selector$2 === 'scala.Float') || ($selector$2 === 'scala.Double')) {
return true;
}
if (true) {
return classNameIsAnyOrAnyVal;
}
})(classFullName);
}
if ($selector$1 === 'boolean') {
return (classNameIsAnyOrAnyVal || (classFullName == 'scala.Boolean'));
}
if ($selector$1 === 'string') {
return (function($selector$3) {
if ($selector$3 === 'scala.Char') {
return self.isChar(anObject);
}
if ($selector$3 === 'scala.String') {
return true;
}
if (true) {
return classNameIsAnyOrAnyRef;
}
})(classFullName);
}
if ($selector$1 === 'function') {
return (function() {
throw new scala.NotImplementedException('Type check of a function isn\'t supported.');
})();
}
if ($selector$1 === 'object') {
if (classNameIsAnyOrAnyRef) {
return true;
}
}
if (true) {
if ((self.classOf(anObject) != null)) {
return self.classOf(anObject).isSubClassOrEqual(classFullName);
}
}
if (true) {
return false;
}
})(goog.typeOf(anObject));

};
s2js.asInstanceOf = function(anObject, className) {
var self = this;
if ((! self.isInstanceOf(anObject, className))) {
(function() {
throw new scala.ClassCastException((((('The object \'' + anObject.toString()) + '\' can\'t be casted to ') + className) + '.'));
})();
}

return anObject;

};
s2js.isUndefined = function(anObject) {
var self = this;
return goog.typeOf(anObject) === 'undefined';};
s2js.isInteger = function(anObject) {
var self = this;
return anObject % 1 === 0;};
s2js.isChar = function(anObject) {
var self = this;
return anObject.length === 1;};
s2js.__class__ = new s2js.Class('s2js', []);
